src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
This will only ever lead to errors, right? I believe we should only use `Data.ServiceProvider`. If that's `null` then we ought to allow the application to crash.